File: compiler/src/feedback/items/FormatFeedback.swift
func registerFormatFeedback(into context: FeedbackContext) {
    context.register(
        id: "missing-format-inputs",
        query: Queries.fact(MissingFormatInputs.self),
        rank: { _, _ in .syntax },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, node, missingInputs in
        writer.write(node)
        writer.write(" needs ")
        writer.write(missingInputs.count, singular: "more input", plural: "more inputs")
        writer.write(".")
        writer.writeBreak()
        writer.write("Try adding code after the string.")
    }

    context.register(
        id: "extra-format-input",
        query: Queries.fact(ExtraFormatInput.self),
        rank: { _, _ in .syntax },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, node, _ in
        writer.write(node)
        writer.write(" isn't used in the format string.")
        writer.writeBreak()
        writer.write("Try removing this input or add another ")
        writer.write(code: "_")
        writer.write(" placeholder.")
    }
}

File: compiler/feedback/src/items/format_feedback.rs
use crate::{FeedbackCtx, FeedbackRank};
use wipple_queries::fact;
use wipple_syntax::expressions::format_expression::{ExtraFormatInput, MissingFormatInputs};

pub fn register(ctx: &mut FeedbackCtx<'_>) {
    ctx.feedback("missing-format-inputs")
        .query(fact::<MissingFormatInputs>)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, node, MissingFormatInputs(count)| {
            writer.node(node);
            writer.string(" needs ");
            writer.singular_plural(*count, "more input", "more inputs");
            writer.string(".");
            writer.line_break();
            writer.string("Try adding code after the string.");
        })
        .register();

    ctx.feedback("extra-format-input")
        .query(fact::<ExtraFormatInput>)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, node, _| {
            writer.node(node);
            writer.string(" isn't used in the format string.");
            writer.line_break();
            writer.string("Try removing this input or add another ");
            writer.code("_");
            writer.string(" placeholder.");
        })
        .register();
}
